widget/textgrid.go

```diff
@@ -0,0 +1,384 @@
+// The grid will use default text style and any previous content and style will be removed.
+func (t *TextGrid) SetText(text string) {
+	var rows [][]TextGridCell
+	for _, line := range strings.Split(text, "\n") {
+		var row []TextGridCell
+		for _, r := range line {
+			row = append(row, TextGridCell{Rune: r})
+		}
+		rows = append(rows, row)
+	}
+
+	t.Content = rows
+	t.Refresh()
+}
+
+// Text returns the contents of the buffer as a single string (with no style information).
+// It reconstructs the lines by joining with a `\n` character.
+func (t *TextGrid) Text() string {
+	var runes []rune
+	for i, row := range t.Content {
+		for _, r := range row {
+			runes = append(runes, r.Rune)
+		}
+
+		if i < len(t.Content)-1 {
+			runes = append(runes, '\n')
+		}
+	}
+
+	return string(runes)
+}
+
+// Row returns the content of a specified row as a slice of TextGridCells.
+// If the index is out of bounds it returns an empty slice.
+func (t *TextGrid) Row(row int) []TextGridCell {
+	if row < 0 || row >= len(t.Content) {
+		return []TextGridCell{}
+	}
+
+	return t.Content[row]
+}
+
+// SetRow updates the specified row of the grid's contents using the specified cell content and style and then refreshes.
+// If the row is beyond the end of the current buffer it will be expanded.
+func (t *TextGrid) SetRow(row int, content []TextGridCell) {
+	if row < 0 {
+		return
+	}
+	for len(t.Content) <= row {
+		t.Content = append(t.Content, []TextGridCell{})
+	}
+
+	t.Content[row] = content
+	t.Refresh()
+}
+
+// SetStyle sets a grid style to the cell at named row and column
+func (t *TextGrid) SetStyle(row, col int, style TextGridStyle) {
+	if row < 0 || col < 0 {
+		return
+	}
+	for len(t.Content) <= row {
+		t.Content = append(t.Content, []TextGridCell{})
+	}
+	content := t.Content[row]
+
+	for len(content) <= col {
+		content = append(content, TextGridCell{})
+	}
+	content[col].Style = style
+}
+
+// SetStyleRange sets a grid style to all the cells between the start row and column through to the end row and column.
+func (t *TextGrid) SetStyleRange(startRow, startCol, endRow, endCol int, style TextGridStyle) {
+	if startRow == endRow {
+		for col := startCol; col <= endCol; col++ {
+			t.SetStyle(startRow, col, style)
+		}
+		return
+	}
+
+	// first row
+	for col := startCol; col < len(t.Content[startRow]); col++ {
+		t.SetStyle(startRow, col, style)
+	}
+
+	// possible middle rows
+	for rowNum := startRow + 1; rowNum < endRow-1; rowNum++ {
+		for col := 0; col < len(t.Content[rowNum]); col++ {
+			t.SetStyle(rowNum, col, style)
+		}
+	}
+
+	// last row
+	for col := 0; col <= endCol; col++ {
+		t.SetStyle(endRow, col, style)
+	}
+}
+
+// CreateRenderer is a private method to Fyne which links this widget to it's renderer
+func (t *TextGrid) CreateRenderer() fyne.WidgetRenderer {
+	t.ExtendBaseWidget(t)
+	render := &textGridRenderer{text: t}
+
+	cell := canvas.NewText("M", color.White)
+	cell.TextStyle.Monospace = true
+	render.cellSize = cell.MinSize()
```

Comment:
```go
render.cellSize = fyne.MeasureText("M", theme.TextSize(), fyne.TextStyle{Monospace:true})
```

Reply:
Thanks, this was not an option when I first coded it up. In now :)